File: PHP/potion-shop-laravel-k8s/app/Http/Controllers/ShopController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Session;

class ShopController extends Controller
{
    /**
     * Purchase messages for random selection
     */
    private const PURCHASE_MESSAGES = [
        "You got a Red Potion! Your hearts are refilled!",
        "A fine purchase, brave hero! May it serve you well!",
        "One Red Potion coming right up! Stay safe out there!",
        "Excellent choice! This potion has saved many adventurers!",
        "A wise investment! You'll thank yourself later!",
        "Thank you for your patronage, noble warrior!",
        "It's dangerous to go alone! Take this potion!",
        "May this potion guide you to victory!",
    ];

    /**
     * Display the shop page
     */
    public function index()
    {
        return view('shop');
    }

    /**
     * Purchase a potion
     */
    public function buy(Request $request)
    {
        // Get current stats from session (default to 0)
        $potionsSold = Session::get('potions_sold', 0);
        $goldEarned = Session::get('gold_earned', 0);

        // Increment stats
        $potionsSold++;
        $goldEarned += 50;

        // Save to session
        Session::put('potions_sold', $potionsSold);
        Session::put('gold_earned', $goldEarned);

        // Get random message
        $message = self::PURCHASE_MESSAGES[array_rand(self::PURCHASE_MESSAGES)];

        return response()->json([
            'success' => true,
            'message' => $message,
            'stats' => [
                'potions_sold' => $potionsSold,
                'gold_earned' => $goldEarned,
            ],
        ]);
    }

    /**
     * Get shop statistics
     */
    public function stats(Request $request)
    {
        return response()->json([
            'potions_sold' => Session::get('potions_sold', 0),
            'gold_earned' => Session::get('gold_earned', 0),
        ]);
    }

    /**
     * Reset shop statistics
     */
    public function reset(Request $request)
    {
        Session::put('potions_sold', 0);
        Session::put('gold_earned', 0);

        return response()->json([
            'success' => true,
            'stats' => [
                'potions_sold' => 0,
                'gold_earned' => 0,
            ],
        ]);
    }

    /**
     * Health check endpoint
     */
    public function health()
    {
        return response()->json([
            'status' => 'ok',
        ]);
    }
}

File: Ruby/potion-shop-rails-k8s/app/controllers/api/shop_controller.rb
class Api::ShopController < ApplicationController
  skip_before_action :verify_authenticity_token

  POTION_COST = 50
  PURCHASE_MESSAGES = [
    "It's dangerous to go alone! Take this.",
    "You got a Red Potion! Your health is restored!",
    "This will help you on your quest, traveler.",
    "A fine choice! That'll be 50 gold.",
    "Thank you for your business, hero!",
    "May this potion serve you well in battle!",
    "One Red Potion, coming right up!",
    "You look like you could use this, adventurer."
  ]

  def buy
    stat = ShopStat.current_stats
    stat.add_purchase(POTION_COST)

    message = PURCHASE_MESSAGES.sample

    render json: {
      message: message,
      potions_sold: stat.potions_sold,
      gold_earned: stat.gold_earned
    }
  end

  def stats
    stat = ShopStat.current_stats

    render json: {
      potions_sold: stat.potions_sold,
      gold_earned: stat.gold_earned
    }
  end

  def reset
    stat = ShopStat.current_stats
    stat.reset_stats

    render json: {
      message: "Shop statistics have been reset!",
      potions_sold: stat.potions_sold,
      gold_earned: stat.gold_earned
    }
  end

  def health
    render json: {
      status: "ok",
      timestamp: Time.current.iso8601,
      service: "potion-shop-rails-k8s"
    }
  end
end
